sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs

```diff
@@ -456,41 +494,54 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
+                // Determine the transport producer to delegate the send operation to.  Because sending to a specific
+                // partition requires a dedicated client, use (or create) that client if a partition was specified.  Otherwise
+                // the default gateway producer can be used to request automatic routing from the Event Hubs service gateway.
 
-            if (String.IsNullOrEmpty(eventBatch.SendOptions.PartitionId))
-            {
-                activeProducer = EventHubProducer;
-            }
-            else
-            {
-                // This assertion is intended as an additional check, not as a guarantee.  There still exists a benign
-                // race condition where a transport producer may be created after the client has been closed; in this case
-                // the transport producer will be force-closed with the associated connection or, worst case, will close once
-                // its idle timeout period elapses.
+                TransportProducer activeProducer;
 
-                Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));
-                activeProducer = PartitionProducers.GetOrAdd(eventBatch.SendOptions.PartitionId, id => Connection.CreateTransportProducer(id, RetryPolicy));
-            }
+                if (String.IsNullOrEmpty(eventBatch.SendOptions.PartitionId))
+                {
+                    activeProducer = EventHubProducer;
+                }
+                else
+                {
+                    // This assertion is intended as an additional check, not as a guarantee.  There still exists a benign
+                    // race condition where a transport producer may be created after the client has been closed; in this case
+                    // the transport producer will be force-closed with the associated connection or, worst case, will close once
+                    // its idle timeout period elapses.
 
-            using DiagnosticScope scope = CreateDiagnosticScope();
+                    Argument.AssertNotClosed(IsClosed, nameof(EventHubProducerClient));
+                    await using var pooledItem = PartitionProducerPool.GetPartitionProducer(eventBatch.SendOptions.PartitionId, Connection, RetryPolicy);
 
-            try
-            {
-                await activeProducer.SendAsync(eventBatch, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception ex)
-            {
-                scope.Failed(ex);
-                throw;
+                    activeProducer = pooledItem.Item;
+                }
+
+                using DiagnosticScope scope = CreateDiagnosticScope();
+
+                try
+                {
+                    await activeProducer.SendAsync(eventBatch, cancellationToken).ConfigureAwait(false);
+                }
+                catch (EventHubsException eventHubException) when (IsPartitionProducerExpired(eventHubException, eventBatch.SendOptions.PartitionId))
```

Comment:
Nice!   This keeps the retry scope to exactly out scenario, since the transport producer is responsible for general retries.